### ISSUE TITLE
Create a new WebExtension API/permission: Provide API to sites

### ISSUE THREAD

**Issue by danfinlay** [brave/brave-browser#643]
At [MetaMask](https://github.com/MetaMask/metamask-plugin), we use the WebExtension "read and write data on all sites you visit" permission for the unusual purpose of providing an additional API to sites a user visits.

A downside of this is that security-conscious users are nervous about approving such all-encompassing permissions, and in fact if our distribution system were compromised, the potential impacts could be large.

Long term, we probably need to be simply integrated into browsers. But _medium term_, we would greatly benefit from a new WebExtension API & accompanying permission that explicitly only provides an API to sites.

Such a permission would *not*:
- Allow us to freely read from websites a user visits
- Run keystroke logging on sites a user visits
- Manipulate how a page otherwise loads for a user.

This permission would *only*:
- Allow us to define an object that could be exposed globally on the `window` object.
- Allow this object to call back when its methods were called.
- Allow this object to pass methods back to the background process before responding.
- Allow registering listeners for events triggered from the background process.

**Comment by BrendanEich** [comment]
Dan, thanks for filing. Has anyone tried getting Google's attention on such an API for extensions being added? Thanks.

**Comment by danfinlay** [comment]
No, I don't think we've filed for Google's attention on this yet, although we did have some discussions with Mozilla a while back:

https://bugzilla.mozilla.org/show_bug.cgi?id=1319168

**Comment by danfinlay** [comment]
I actually am just seeing now that `externally_connectable` from websites got merged.. I need to get reacquainted with that proposal to see how much of this that does.

**Comment by danfinlay** [comment]
So that issue is not perfectly aligned with this one, only allows a port to be created between sites and a specific (known ID) extension, not a generic browser API that any extension can provide.

**Comment by danfinlay** [comment]
Another discussion on this topic is here: https://github.com/MetaMask/metamask-extension/issues/813#issuecomment-261363677

We thought maybe the `Component.utils.cloneInto` api could be used for this with some adjustments.

**Comment by danfinlay** [comment]
Actually maybe we need to revisit this on the MetaMask side. Looks like the `externally_connectable` from sites feature was added so that extensions like MetaMask could do this.

Here's the Chromium discussion on that topic:
https://bugs.chromium.org/p/chromium/issues/detail?id=679238

**Comment by guest271314** [comment]
If this has not been solved yet, note that it is possible to dynamically set `externally_connectable` property values of a Chromium app.

**Comment by guest271314** [comment]
```
(async(connectable) => {
  const dir = await self.showDirectoryPicker(); // select app folder
  const status = await dir.requestPermission({writable: true});
  const fileHandle = await dir.getFileHandle("manifest.json", {create: false});
  const file = await fileHandle.getFile();
  const manifest = await file.text();
  const text = manifest.replace(/"matches": \[.*\]/, `"matches":${connectable}`);
  const writer = await fileHandle.createWritable({keepExistingData:false});
  await writer.write(text); 
  await writer.close();
})(JSON.stringify([`${location.origin}/*`]));
```

**Comment by bsclifton** [comment]
This issue hasn't had an update in a while - it's met the criteria for being stale. The issue is going to be closed for now. If this is still a problem, please leave a comment or re-open the issue. Thanks!

**Comment by guest271314** [comment]
@danfinlay

> Actually maybe we need to revisit this on the MetaMask side. Looks like the `externally_connectable` from sites feature was added so that extensions like MetaMask could do this.
> 
> Here's the Chromium discussion on that topic: https://bugs.chromium.org/p/chromium/issues/detail?id=679238

Yeah, `externally_connectable` works on all domains with something like 

```
"externally_connectable": {
    "matches": [
      "<all_urls>"
    ],
    "ids": [
      "*"
    ]
  },
```

Now there's structured cloning, too for extension messaging; instead of just JSON, with 

```
  "message_serialization": "structured_clone",
```
